File: src/lexer.zig
const std = @import("std");
const alloc = @import("allocator.zig");
const print = std.debug.print;
const char = u8;

pub const TokenType = enum(i32) {
    TT_EQUALITY,
    TT_LT,
    TT_GT,
    TT_GTE,
    TT_LTE,
    TT_ADD,
    TT_SUB,
    TT_MUL,
    TT_DIV,
    TT_EXP,

    TT_NUM,
    TT_IDENT,
    TT_FLOAT,
    TT_KEYWORD,
    TT_NEWLINE,
    TT_ASSIGN,
    TT_LPAREN,
    TT_RPAREN,
    TT_LBRACE,
    TT_RBRACE,
    TT_LBRACKET,
    TT_RBRACKET,
    TT_COMMA,
    TT_FUNC_RETURN_TYPE,
    TT_INCLUDE,
    TT_STRING,
    TT_EOF,
    TT_SPACE,
    TT_STAR,
    TT_ERROR,
    TT_BINDING,
    TT_RANGE,
    TT_COMPILER_CREATED,
    TT_BIT_AND,
    TT_BIT_OR,
    TT_DECREMENT,
    TT_INCREMENT,
    TT_AND,
    TT_OR,
    TT_FIELD_ACCESSOR, // 'struct_name.value'
    TT_POUND,
    TT_COLON, // used for denoting array type length, ex: `ages [int:5] = ...`
    TT_VARARG,

    pub fn op_precedence(self: TokenType) i32 {
        switch (self) {
            .TT_EQUALITY => 0,
            .TT_LT => 1,
            .TT_GT => 1,
            .TT_GTE => 2,
            .TT_LTE => 2,
            .TT_ADD => 3,
            .TT_SUB => 3,
            .TT_MUL => 4,
            .TT_DIV => 4,
            .TT_EXP => 5,
            else => @panic(print("TokenType: {d} has no precedence!", .{self})),
        }
    }
};

pub fn Token(comptime lex_size: i32, comptime file_size: i32) type {
    return struct {
        token_type: TokenType,
        lexeme: [lex_size]u8,
        linenum: i32,
        linedepth: i32,
        filename: [file_size]u8,
    };
}

pub fn NewToken(lex_size: i32, filename_size: i32, token_type: TokenType, lexeme: []u8, filename: []u8, linenum: i32) Token {
    return Token(lex_size, filename_size){
        .token_type = token_type,
        .lexeme = lexeme,
        .linenum = linenum,
        .linedepth = 0,
        .filename = filename,
    };
}

pub fn Lexer() type {
    return struct {
        buffer: std.ArrayList(char),
        tokens: std.ArrayList(Token),
        linenum: i32,
        linedepth: i32,
        current_char: char,
        last_token: Token,
        filename: []const u8,

        fn lexer_make_token(self: *Lexer, token_type: TokenType, lexeme: []u8) *Token {
            const point = try alloc.create(Token(lexeme.len, self.filename.len));
            defer alloc.destroy(point);
            point.* = NewToken(lexeme.len, self.filename.len, token_type, lexeme, self.filename, self.linenum);
            return point;
        }

        fn check_for_string(self: *Lexer) *Token {
            var lexeme = std.ArrayList(char);
            if (self.current_char == '"') {
                try lexeme.append('"');
                self.current_char = self.buffer.pop();
                self.linedepth += 1;

                while (self.current_char != '"') {
                    try lexeme.append(self.current_char);
                    self.current_char = self.buffer.pop();
                    self.linedepth += 1;
                }

                try lexeme.append('"');
                return self.lexer_make_token(.TT_STRING, lexeme.items);
            }
            return null;
        }

        fn followed_by(self: *Lexer, expected_char: u8, expected_type: TokenType, expected_lexeme: []const u8) *Token {
            self.current_char = self.buffer.pop();
            self.linedepth += 1;
            if (self.current_char == expected_char) {
                return self.lexer_make_token(expected_type, expected_lexeme);
            }
            return null;
        }

        fn symbol_handle_dot(self: *Lexer) *Token {
            const first_peek = self.buffer.pop();
            self.current_char = first_peek;

            if ((self.last_token.token_type == .TT_IDENT) and (std.ascii.isAlphabetic(first_peek) or first_peek == '_')) {
                try self.buffer.insert(first_peek);
                return self.lexer_make_token(.TT_FIELD_ACCESSOR, ".");
            }

            if (first_peek == '.') {
                const second_peek = self.buffer.pop();
                if (second_peek == '.') {
                    const tok_type = .TT_RANGE;
                    if (self.last_token.token_type == .TT_IDENT) {
                        tok_type = .TT_VARARG;
                    }

                    return self.lexer_make_token(tok_type, "...");
                }
            }

            return null;
        }

        fn lex_double_symbol(self: *Lexer, default_type: TokenType, expected_char: u8, expected_type: TokenType, expected_lexeme: []u8) *Token {
            const ret = self.followed_by(expected_char, expected_type, expected_lexeme);
            if (ret == null) {
                return ret;
            }

            return self.lexer_make_token(default_type, expected_char);
        }

        fn lex_for_symbol(self: *Lexer) *Token {
            const retval: *Token = self.check_for_string();
            if (retval == null) {
                return null;
            }

            var symbols = std.AutoHashMap(u8, i32).init(alloc);
            defer symbols.deinit();
            comptime {
                symbols.put('(', .TT_LPAREN);
                symbols.put(')', .TT_RPAREN);
                symbols.put('*', .TT_MUL);
                symbols.put('/', .TT_DIV);
                symbols.put(',', .TT_COMMA);
                symbols.put('[', .TT_LBRACKET);
                symbols.put(']', .TT_RBRACKET);
                symbols.put('{', .TT_LBRACE);
                symbols.put('}', .TT_RBRACE);
                symbols.put('#', .TT_POUND);
            }

            switch (self.current_char) {
                ':' => return self.followed_by(':', .TT_BINDING, "::"),
                '-' => {
                    const peekahead = self.buffer.pop();
                    if (peekahead == '>') {
                        return self.lexer_make_token(.TT_FUNC_RETURN_TYPE, "->");
                    } else if (peekahead == '-') {
                        return self.lexer_make_token(.TT_DECREMENT, "--");
                    }
                    return self.lexer_make_token(.TT_SUB, "-");
                },
                '.' => return self.symbol_handle_dot(),
                '=' => self.lex_double_symbol(.TT_ASSIGN, '=', .TT_EQUALITY, "=="),
                '+' => self.lex_double_symbol(.TT_ADD, '+', .TT_INCREMENT, "++"),
                '>' => self.lex_double_symbol(.TT_GT, '>', .TT_GTE, ">="),
                '<' => self.lex_double_symbol(.TT_LT, '=', .TT_LTE, "<="),
                '&' => self.lex_double_symbol(.TT_BIT_AND, '&', .TT_AND, "&&"),
                '|' => self.lex_double_symbol(.TT_BIT_OR, '|', .TT_OR, "||"),
                else => {
                    const exists = symbols.contains(self.current_char);
                    if (!exists) {
                        return null;
                    }
                    const tok_type = symbols[self.current_char];

                    return self.lexer_make_token(tok_type, self.current_char);
                },
            }
        }

        fn lex_for_numbers(self: *Lexer) *Token {
            if (!std.ascii.isDigit(self.current_char)) {
                return null;
            }

            var dot_count = 0;
            var lexeme = std.ArrayList(char);
            var tok_type = .TT_NUM;
            while (std.ascii.isDigit(self.current_char) or (self.current_char == '.' and dot_count == 0)) {
                if (self.current_char == '.') {
                    tok_type = .TT_FLOAT;
                    dot_count += 1;
                }

                try lexeme.append(self.current_char);
                self.current_char = self.buffer.pop();
                self.linedepth += 1;
            }

            if (lexeme.getLast() == '.') {
                tok_type = .TT_NUM;
                lexeme.pop();
                self.buffer.insert(0, '.');
            }

            self.buffer.insert(0, self.current_char);
            const token = self.lexer_make_token(tok_type, lexeme.items);
            return token;
        }

        fn lex_for_identifiers(self: *Lexer) *Token {
            if (!std.ascii.isAlphabetic(self.current_char) and self.current_char != '_') {
                return null;
            }

            var lexeme = std.ArrayList(char);
            const keywords = std.AutoHashMap([]const u8, u8);
            defer keywords.deinit();
            comptime {
                keywords.put("int", 0);
                keywords.put("char", 0);
                keywords.put("void", 0);
                keywords.put("i16", 0);
                keywords.put("i32", 0);
                keywords.put("i64", 0);
                keywords.put("f32", 0);
                keywords.put("f64", 0);
                keywords.put("bool", 0);
                keywords.put("include", 0);
                keywords.put("for", 0);
                keywords.put("while", 0);
                keywords.put("in", 0);
                keywords.put("if", 0);
                keywords.put("else", 0);
                keywords.put("break", 0);
                keywords.put("continue", 0);
                keywords.put("fallthrough", 0);
                keywords.put("ret", 0);
                keywords.put("struct", 0);
            }

            while (std.ascii.isAlphabetic(self.current_char) or std.ascii.isDigit(self.current_char) or self.current_char == '_') {
                try lexeme.append(self.current_char);
                self.current_char = self.buffer.pop();
                self.linedepth += 1;
            }

            self.buffer.insert(0, self.current_char);

            const token = self.lexer_make_tokens(.TT_IDENT, lexeme.items);
            if (!keywords.contains(lexeme.items)) {
                token.token_type = .TT_KEYWORD;
            }

            return token;
        }

        fn unget_token(self: *Lexer) void {
            self.tokens.insert(0, self.last_token);
        }
    };
}

File: src/lexer.cpp
#include <string>
#include <cctype>
#include <unordered_map>
#include <fstream>
#include "../include/error_logger.h"
#include "../include/lexer.h"
#include "../include/token.h"

using namespace std;

SageLexer::SageLexer(istream *stream, string sourcename): char_buffer(stream), sourcename(sourcename) {
    linenum = 1;
    linedepth = 0;
    current_token = new Token();
    last_token = Token();
    peeked_tokens = stack<Token>();
}

SageLexer::SageLexer() : char_buffer(nullptr) {
    sourcename = "";
    linenum = 0;
    linedepth = 0;
    current_token = new Token();
    last_token = Token();
    peeked_tokens = stack<Token>();
}

SageLexer::~SageLexer() {
    if (current_token != nullptr) {
        delete current_token;
        current_token = nullptr;
    }
}

Token *SageLexer::check_for_string() {
    string lexeme;
    if (current_char == '"') {
        lexeme += '"';
        char_buffer->get(current_char);
        linedepth++;

        while (current_char != '"' && current_char != '\n' && !char_buffer->eof()) {
            lexeme += current_char;
            char_buffer->get(current_char);
            linedepth++;
            if (char_buffer->eof() || char_buffer->fail()) break;
        }

        if (current_char != '"') {
            Token tok = Token(TT_STRING, lexeme, linenum);
            ErrorLogger::get().log_error_unsafe(tok, "Unterminated string literal.", SYNTAX);
        }

        lexeme += '"';
        return lexer_make_token(TT_STRING, lexeme);
    }

    return nullptr;
}

Token *SageLexer::check_for_character_literal() {
    string lexeme;
    if (current_char == '\'') {
        lexeme += '\'';
        char_buffer->get(current_char);
        linedepth++;
        char_buffer->get(current_char);
        linedepth++;

        if (current_char != '\'') {
            Token tok = Token(TT_CHARACTER_LITERAL, "\'", linenum);
            ErrorLogger::get().log_error_unsafe(tok, "Character literals cannot be more than one character long.", SYNTAX);
        }

        char_buffer->get(current_char);
        linedepth++;

        lexeme += '\'';
        return lexer_make_token(TT_CHARACTER_LITERAL, lexeme);
    }

    return nullptr;
}

Token *SageLexer::handle_symbol_case(
    char default_char, TokenType default_type,
    TokenType target_type, string target_symbol
) {
    Token *ret = followed_by(default_char, target_type, target_symbol);
    if (ret != nullptr) {
        return ret;
    }
    return lexer_make_token(default_type, string(1, default_char));
}

Token *SageLexer::lex_for_symbols() {
    Token *return_val = check_for_string();
    if (return_val != nullptr) {
        return return_val;
    }

    return_val = check_for_character_literal();
    if (return_val != nullptr) {
        return return_val;
    }

    unordered_map<char, TokenType> SYMBOLS = {
        {'(', TT_LPAREN},
        {')', TT_RPAREN},
        {'*', TT_MUL},
        {'/', TT_DIV},
        {'%', TT_MODULO},
        {',', TT_COMMA},
        {'[', TT_LBRACKET},
        {']', TT_RBRACKET},
        {'{', TT_LBRACE},
        {'}', TT_RBRACE},
        {'#', TT_POUND},
        {'@', TT_POINTER_DEREFERENCE}
    };

    char peekahead;
    char first_peek;
    switch (current_char) {
        case ':': {
            auto retval = followed_by(':', TT_BINDING, "::");
            if (retval != nullptr) {
                return retval;
            }

            return lexer_make_token(TT_COLON, ":");
        }
        case '^': {
            if (last_token.token_type == TT_NUM || last_token.token_type == TT_FLOAT) {
                return lexer_make_token(TT_EXPONENT, "^");
            }
            return lexer_make_token(TT_POINTER_REFERENCE, "^");
        }

        case '-':
            char_buffer->get(peekahead);
            if (peekahead == '>') {
                linedepth++;
                return lexer_make_token(TT_FUNC_RETURN_TYPE, "->");
            }
            if (peekahead == '-') {
                linedepth++;
                return lexer_make_token(TT_DECREMENT, "--");
            }

            char_buffer->putback(peekahead);
            return lexer_make_token(TT_SUB, "-");

        case '.':
            char_buffer->get(first_peek);
            current_char = first_peek;

            // if the '.' is in between an identifier token and unicode chars then its a field accessor
            // NOTE: isalpha(current_char) returns non zero value if the character is a letter
            if (last_token.token_type == TT_IDENT && (isalpha(current_char) != 0 || current_char == '_')) {
                char_buffer->putback(first_peek);
                return lexer_make_token(TT_FIELD_ACCESSOR, ".");
            }

            if (first_peek == '.') {
                char second_peek;
                char_buffer->get(second_peek);
                if (second_peek == '.') {
                    TokenType tok_type = TT_RANGE;
                    if (last_token.token_type == TT_IDENT) {
                        tok_type = TT_VARARG;
                    }

                    return lexer_make_token(tok_type, "...");
                }
            }

            return nullptr;

        case '=':
            return handle_symbol_case('=', TT_ASSIGN, TT_EQUALITY, "==");

        case '+':
            return handle_symbol_case('+', TT_ADD, TT_INCREMENT, "++");

        case '>':
            return handle_symbol_case('>', TT_GT, TT_GTE, ">=");

        case '<':
            return handle_symbol_case('<', TT_LT, TT_LTE, "<=");

        case '&':
            return handle_symbol_case('&', TT_AND, TT_BIT_AND, "&&");

        case '|':
            return handle_symbol_case('|', TT_OR, TT_BIT_OR, "||");

        default:
            if (SYMBOLS.find(current_char) == SYMBOLS.end()) {
                return nullptr;
            }

            return lexer_make_token(SYMBOLS[current_char], string(1, current_char));
    }
}

Token *SageLexer::followed_by(char expected_char, TokenType expected_type, string expected_lexeme) {
    char_buffer->get(current_char);
    linedepth++;
    if (current_char == expected_char) {
        return lexer_make_token(expected_type, expected_lexeme);
    }

    char_buffer->putback(current_char);
    linedepth--;
    return nullptr;
}

Token *SageLexer::lexer_make_token(TokenType type, string lexeme, int depth) {
    current_token->lexeme = lexeme;
    current_token->token_type = type;
    current_token->filename = this->sourcename;
    current_token->linenum = this->linenum;
    current_token->linedepth = depth == -1 ? this->linedepth : depth;
    return current_token;
}

Token *SageLexer::lex_for_numbers() {
    if (!isdigit(current_char)) {
        return nullptr;
    }

    int dot_count = 0;
    string lexeme = "";
    TokenType tok_type = TT_NUM;
    while ((isdigit(current_char) || (current_char == '.' && dot_count == 0)) && !char_buffer->eof()) {
        if (current_char == '.') {
            tok_type = TT_FLOAT;
            dot_count++;
        }

        lexeme += string(1, current_char);
        char_buffer->get(current_char);
        linedepth++;

        if (char_buffer->eof() || char_buffer->fail()) {
            break;
        }
    }

    // if the number ends in a '.' then this is a number before the range operator and not actually a float number
    // we know this because the loop above handles floating point notation.
    if (lexeme.length() != 0 && lexeme[lexeme.length() - 1] == '.') {
        tok_type = TT_NUM;
        lexeme.pop_back();
        char_buffer->putback('.');
    }

    char_buffer->putback(current_char);
    return lexer_make_token(tok_type, lexeme);
}

Token *SageLexer::lex_for_identifiers() {
    if (!isalpha(current_char) && current_char != '_') {
        return nullptr;
    }

    string lexeme = "";

    set<string> KEYWORDS = {
        "int",
        "char",
        "void",
        "i32",
        "i64",
        "f32",
        "f64",
        "bool",
        "for",
        "while",
        "if",
        "else",
        "break",
        "continue",
        "fallthrough",
        "ret",
        "struct",
        "with", // previously: 'using'
        "run",
        "true",
        "false",
        "compact",
        "defer"
    };

    int starting_linedepth = linedepth;
    while ((isalpha(current_char) || isdigit(current_char) || current_char == '_') && !char_buffer->eof()) {
        lexeme += string(1, current_char);
        char_buffer->get(current_char);
        linedepth++;

        if (char_buffer->eof() || char_buffer->fail()) {
            break;
        }
    }

    char_buffer->putback(current_char);

    lexer_make_token(TT_IDENT, lexeme, starting_linedepth);
    if (KEYWORDS.find(lexeme) != KEYWORDS.end()) {
        current_token->token_type = TT_KEYWORD;
    }

    return current_token;
}

Token *SageLexer::get_token() {
    char_buffer->clear();
    Token *tok;

    // we check that the size is greater than one because the last_token will be held inside the buffer
    if (!peeked_tokens.empty()) {
        Token t = peeked_tokens.top();
        peeked_tokens.pop();

        current_token->fill_with(t);
        return current_token;
    }

    // note: CRITICAL!! We have to call get() before we check eof() becaus eof only updates to eof after a *failed* get() call.
    char_buffer->get(current_char);
    linedepth++;
    while ((current_char == ' ' || current_char == '\t') && !char_buffer->eof()) {
        char_buffer->get(current_char);
        linedepth++;
        if (char_buffer->eof() || char_buffer->fail()) break;
    }

    if (char_buffer->eof()) {
        tok = lexer_make_token(TT_EOF, "eof");
        last_token = *tok;
        return tok;
    }

    if (current_char == '\n') {
        tok = lexer_make_token(TT_NEWLINE, "\n");
        linenum++;
        linedepth = 0;
        last_token = *tok;
        return tok;
    }

    tok = lex_for_symbols();
    if (tok != nullptr) {
        last_token = *tok;
        return tok;
    }

    tok = lex_for_identifiers();
    if (tok != nullptr) {
        last_token = *tok;
        return tok;
    }

    tok = lex_for_numbers();
    if (tok != nullptr) {
        last_token = *tok;
        return tok;
    }

    string undefined_lexeme = str(current_char);
    tok = lexer_make_token(TT_ERROR, "unrecognized symbol");
    ErrorLogger::get().log_error_unsafe(*tok, sen("unrecognized symbol:", undefined_lexeme), SYNTAX);
    last_token = *tok;
    return tok;
}

void SageLexer::unget_token() {
    peeked_tokens.push(last_token);
}
